File: bayesmendel-comparator/r/R/risk_api_client.R
# Call Evagene's /risk/calculate endpoint.  The model name is data.

HTTP_OK_LOWER <- 200L
HTTP_OK_UPPER <- 300L

#' Raise an api_error condition.
#' @noRd
stop_api <- function(message) {
  stop(structure(
    list(message = message, call = sys.call(-1)),
    class = c("api_error", "error", "condition")
  ))
}

#' Run a single Evagene risk calculation.
#'
#' @param gateway     A gateway function (see `make_httr2_gateway()`).
#' @param base_url    Evagene base URL (trailing slash optional).
#' @param api_key     Value for the `X-API-Key` header.
#' @param pedigree_id UUID string.
#' @param model       Model name (e.g. "BRCAPRO", "MMRpro", "PancPRO").
#' @param counselee_id Optional UUID string.
#' @return A named list parsed from the response JSON.
calculate_risk <- function(gateway, base_url, api_key,
                           pedigree_id, model, counselee_id = NULL) {
  url <- sprintf(
    "%s/api/pedigrees/%s/risk/calculate",
    sub("/$", "", base_url), pedigree_id
  )
  body <- list(model = model)
  if (!is.null(counselee_id)) {
    body$counselee_id <- counselee_id
  }

  response <- gateway(list(
    url = url,
    headers = c("X-API-Key" = api_key, "Accept" = "application/json"),
    body_json = jsonlite::toJSON(body, auto_unbox = TRUE)
  ))

  if (response$status < HTTP_OK_LOWER || response$status >= HTTP_OK_UPPER) {
    stop_api(sprintf(
      "Evagene API returned HTTP %d for model %s", response$status, model
    ))
  }

  parsed <- tryCatch(
    jsonlite::fromJSON(response$json_body, simplifyVector = FALSE),
    error = function(e) {
      stop_api(sprintf(
        "Evagene API returned invalid JSON for model %s: %s",
        model, conditionMessage(e)
      ))
    }
  )
  if (!is.list(parsed) || is.null(names(parsed))) {
    stop_api(sprintf("Evagene API returned non-object JSON for model %s", model))
  }
  parsed
}

File: nice-traffic-light/python/src/nice_traffic_light/risk_api_client.py
"""Call Evagene's ``risk/calculate`` endpoint for the NICE model."""

from __future__ import annotations

from typing import Any

from .http_gateway import HttpGateway

_HTTP_OK_LOWER = 200
_HTTP_OK_UPPER = 300


class ApiError(RuntimeError):
    """Raised when the Evagene API is unreachable or returns a non-2xx response."""


class RiskApiClient:
    def __init__(
        self,
        *,
        base_url: str,
        api_key: str,
        http: HttpGateway,
    ) -> None:
        self._base_url = base_url.rstrip("/")
        self._api_key = api_key
        self._http = http

    def calculate_nice(
        self,
        pedigree_id: str,
        *,
        counselee_id: str | None = None,
    ) -> dict[str, Any]:
        url = f"{self._base_url}/api/pedigrees/{pedigree_id}/risk/calculate"
        body: dict[str, Any] = {"model": "NICE"}
        if counselee_id is not None:
            body["counselee_id"] = counselee_id

        response = self._http.post_json(url, headers=self._headers(), body=body)
        if not _HTTP_OK_LOWER <= response.status_code < _HTTP_OK_UPPER:
            raise ApiError(f"Evagene API returned HTTP {response.status_code} for {url}")

        payload = response.json()
        if not isinstance(payload, dict):
            raise ApiError(f"Evagene API returned non-object JSON: {type(payload).__name__}")
        return payload

    def _headers(self) -> dict[str, str]:
        return {
            "X-API-Key": self._api_key,
            "Content-Type": "application/json",
            "Accept": "application/json",
        }
